src/app/donation-start/donation-start-form/donation-start-form.component.ts

```diff
@@ -689,6 +696,68 @@ export class DonationStartFormComponent implements AfterContentChecked, AfterCon
     }
   }
 
+  private runFinalPreSubmitUpdate() {
+    // Even if next guard fails, we want to prevent attempting to pay.
+    this.runningFinalPreSubmitUpdate = true;
+
+    // todo combine with below, improve Matomo logging
+    if (!this.donation) {
+      const errorCodeDetail = '[code A1]'; // Donation property absent.
+
+      const errorMessage = `Missing donation information – please refresh and try again, or email hello@biggive.org quoting ${errorCodeDetail} if this problem persists`;
+
+      this.toast.showError(errorMessage);
+
+      this.stripeError = errorMessage;
+
+      this.stripeResponseErrorCode = undefined;
+      this.matomoTracker.trackEvent(
+        'donate_error',
+        'submit_missing_donation_basics',
+        `Donation not set or form invalid ${errorCodeDetail}`,
+      );
+      return;
+    }
+
+    if (!this.donation || !this.campaign || !this.campaign.charity.id || !this.psp) {
+      this.donationUpdateError = true;
+      // todo log to Matomo
+      this.toast.showError("Sorry, we can't submit your donation right now.");
+      return;
+    }
+
+    if (this.paymentGroup.value.billingPostcode) {
+      this.donation.billingPostalAddress = this.paymentGroup.value.billingPostcode;
+      this.donation.countryCode = this.paymentGroup.value.billingCountry;
+      this.donationService.updateLocalDonation(this.donation);
+    }
```

Comment:
With this in its later position I got a MatchBot validation / assertion error on billing postcode; can't remember now if on Update or Confirm.

Reply:
@NoelLH to clarify was that an expected error when you provided an invalid postcode? Or an error that's a problem? I'm assuming the former.

Reply:
In its previous position in the code with some other updates, it was an unexpected error. The value was not in the payload sent to MatchBot even though it was on the page. I lost the thread as to exactly why the move was needed and have little memory of it now.

Moved here, so far it seems to work.